### PR TITLE
Add recipe for AllenNLP (do not merge yet)

### PR DESCRIPTION
**Pull request by nelson-liu**
This PR adds a recipe for [AllenNLP](http://allennlp.org/), a natural language processing research library built on top of PyTorch.

This isn't quite ready for merging yet --- I will ping back and update the description when it is. I'm planning on waiting for their next PyPI release, after which I'll verify that this package still works with the `sdist` tarball and then request reviews.

cc: @schmmd

### CODE REVIEW

**Comment by nelson-liu** [comment]
@conda-forge/core I'm a bit confused why circleci and travis say "Computed that there are 0 distributions to build from 1 recipes".

It seems like the build shouldn't be skipped since skip is true for only windows builds or with python versions earlier than 3.6 (`# [win or py<36]`), while both travis and circle note that the python version is 3.6. any insight?

**Comment by scopatz** [comment]
isn't the `py` variable in selectors a string?  shouldn't it be `py >= '36'`

**Comment by nelson-liu** [comment]
@scopatz I thought it was too, but running `conda build` gives me:

```
Error: Invalid selector in meta.yaml line 19:
offending line:
  skip: True  # [win or py>='36']
exception:
'>=' not supported between instances of 'int' and 'str'
```

Perhaps [the docs](https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#preprocessing-selectors) should be updated for this?

**Comment by scopatz** [comment]
Woah, that is an issue.  I am suprised it is a string.  That means those docs are wildly wrong.  Please open up an issue at https://github.com/conda/conda-build 

CC @msarahan @kalefranz @jjhelmus

File: recipes/allennlp/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  skip: True  # [win or py>=36]
+  number: 0
+  script: python -m pip install --no-deps --ignore-installed .
+
+requirements:
+  build:
+    - pip
+    - python >=3.6
+  run:
+    - python >=3.6
```
Comment (jakirkham) [inline recipes/allennlp/meta.yaml]: Would remove the explicit versions from the `python`s here. `skip` should be sufficient to get the intended behavior. These will likely complicate matters.

File: recipes/allennlp/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+{% set hash_value = "dbc0b7de268a14ad978ad29ceb785c0a22cab13e2786f04fe933b4c71e9ad7aa" %}
+
+package:
+  name: '{{ name|lower }}'
+  version: '{{ version }}'
+
+source:
+  fn: '{{ name }}-{{ version }}.{{ file_ext }}'
+  # Fix this when the new release is officially out.
+  url: https://github.com/allenai/allennlp/archive/v0.4.3.tar.gz
```
Comment (jakirkham) [inline recipes/allennlp/meta.yaml]: Would template version in this URL to avoid accidentally publishing something later where the URL and version point to different things.

File: recipes/allennlp/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    - psycopg2
+    - requests >=2.18
+    - tqdm >=4.19
+    - editdistance
+    - h5py
+    - scikit-learn
+    - scipy
+    - pytz ==2017.3
+    - unidecode
+test:
```
Comment (jakirkham) [inline recipes/allennlp/meta.yaml]: Would space out the sections for readability.

**Comment by nelson-liu** [comment]
@jakirkham looks like circleci is running python 3.5 —- does this mean the >=3.6 requirement on python at run is necessary?

File: recipes/allennlp/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+source:
+  fn: '{{ name }}-{{ version }}.{{ file_ext }}'
+  # Fix this when the new release is officially out.
+  url: https://github.com/allenai/allennlp/archive/v{{ version }}.tar.gz
+  # url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  '{{ hash_type }}': '{{ hash_value }}'
+
+build:
+  skip: True  # [win or py>=36]
```
Comment (isuruf) [inline recipes/allennlp/meta.yaml]: Why did you change this to `py>=36`? It was correct when it was `py<36`

Comment (nelson-liu) [reply recipes/allennlp/meta.yaml]: Ah, seeing as it is 3.6 only I figured they’d be more or less the same. I will change it back, thanks.

Comment (isuruf) [reply recipes/allennlp/meta.yaml]: They are not the same. They are opposites of each other.

`py<36` skips python 2.7, 3.5 and runs 3.6
`py>=36` skips python 3.6 and runs 2.7, 3.5

Comment (nelson-liu) [reply recipes/allennlp/meta.yaml]: Oh whoops, you’re right, thanks for pointing that out. That probably explains the issue I’m having with ci

**Comment by nelson-liu** [comment]
ok, this seems to work well for linux! Thanks for everyones' help!

However, the package doesn't work for osx since the `pytorch` in the default channel doesn't support `osx-64`. Note that the `pytorch` package in the `pytorch` channel itself supports windows, osx, and linux; allennlp currently works with only osx and linux, but we're moving towards windows support as well since pytorch also supports windows as of their latest release. It'd be nice if it was easy to build conda packages for all these platforms when the code is windows-ready.

What are the options here? I'm fine with having users install pytorch separately, but I'd like to run the `osx-64` tests on CI if possible... is it possible to use a package from a different channel for the tests / build? 

I'm planning on recommending that users install allennlp with something like `conda install pytorch allennlp -c pytorch -c conda-forge` anyway, so they get the `pytorch` in the `pytorch` channel and pull `allennlp` from `conda-forge`.

again, thanks for all the help thus far, it's much appreciated!

**Comment by jakirkham** [comment]
How do you feel about adding a `pytorch` recipe? ;)

**Comment by nelson-liu** [comment]
@jakirkham I was hoping it wouldn't come to that 🙂 . I took a look at #3780 earlier, and I agree that it needs a newer version.

It also looks like there were some issues there wrt conda-forge and cuda --- what's the status on that?

lastly, could you clarify what you mean by:

> We could skip macOS temporarily and try building it using CircleCI's macOS workers in the feedstock.

What difference does it make if we're using circleci mac workers vs the travis ones? Or are you referring to trying to build pytorch within the circleci macos workers? Sorry, probably being dense.
